examples/self_sustaining_parallel.py

```diff
@@ -0,0 +1,205 @@
+            task_feedback_queue.put(new_task)
+
+            # ...until the stopping criterion is met...
+            if new_task is None:
+                break
+
+        # ...then, retrieve the remaining outputs.
+        for output in _output_generator:
+            task_completion_callback(output)
+
+    def _input_generator(
+            self,
+            starting_tasks,
+            task_from_output_fn,
+            task_feedback_queue
+    ):
+        '''The input generator, that forwards new tasks created in the calling
+        thread to the workers.
+        '''
+        nb_pending_results = 0
+
+        # First, dispatch the starting tasks.
+        for task in starting_tasks:
+            nb_pending_results += 1
+            yield task
+
+        # Need enough starting tasks start the dispatch with respect to the
+        # batch_size and pre_dispatch parameters.
+        if nb_pending_results < (
+                minimum_starting_tasks_nb := (
+                    self.parallel.batch_size *
+                    self.parallel._pre_dispatch_amount
+            )
+        ):
+            raise ValueError(
+                "Expected at least batch_size * parallel._pre_dispatch_amount "
+                f"= {minimum_starting_tasks_nb} starting tasks, but only got "
+                f"{nb_pending_results} starting tasks."
+            )
+
+        while True:
+            # Wait for a next task to be available...
+            new_task = task_feedback_queue.get()
+
+            # ...if the feedback loop does not return a new task this time,
+            # it means the compute stops after the remaining pending tasks
+            # are done.
+            if new_task is None:
+                return
+
+            yield new_task
+
+
+##############################################################################
+# Now let's define our starting tasks, the feedback loop, and a counter
+# callback.
+
+def add_one(i):
+    return i + 1
+
+
+starting_tasks = (joblib.delayed(add_one)(i) for i in range(10))
+
+
+def task_from_output_fn(output):
+    # Stopping criterion: after the first time an integer went above 100, stop
+    # issuing new tasks.
+    if output >= 100:
+        return
+
+    return joblib.delayed(add_one)(output)
+
+
+class CounterCallback:
+    def __init__(self):
+        self.counter = 0
+
+    def __call__(self, output):
+        self.counter += 1
+
+    def get_count(self):
+        return self.counter
+
+
+##############################################################################
+# Start the compute and check the counter.
+
+callback = CounterCallback()
+SelfSustainingParallelWork(parallel_n_jobs=2)(
+    starting_tasks,
+    task_from_output_fn,
+    callback
+)
+
+# NB: in our example, the number of tasks done can change between runs since
+# it depends on concurrency between workers until one of them meets the
+# stopping criterion first.
+print(f"{callback.get_count()} tasks done.")
```

Comment:
Should we implement verbosity in the class to avoid having to add a callback in the example and thus making it simpler (I always try to make examples as simple as possible)

Reply:
Agree that the usability of the class can be refined. Other points to note with this first draft:

- the callback feature is actually not required since the user could directly add callbacks inside the `task_from_output_fn

- if anything the callback should be moved inside the `try/except` block in `__call__`, to avoid a bug where the backend would deadlock if an exception is raised in the callback.